File: extras/Python/samples/arduino_controller/arduino_controller.py
"""ArduinoController — LED blink-frequency control via slider.

Port of C# ``ArduinoController.cs``. The browser shows a slider and a toggle
button. Slider events send ``SetLedFrequency`` commands wrapped in
:class:`CollapseCommandStrategy` to avoid queue lag.

Pairs with ``examples/ArduinoController/ArduinoController.ino``.
"""
from __future__ import annotations

import sys, os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from cmd_messenger import (
    BoardType,
    CmdMessenger,
    CollapseCommandStrategy,
    ReceivedCommand,
    SendCommand,
)
from cmd_messenger.transport.serial import SerialSettings, SerialTransport
from shared import WebForm

PORT = "COM6"
BAUD = 115200


# Command IDs (must match Arduino sketch enum).
class Command:
    ACKNOWLEDGE = 0
    ERROR = 1
    SET_LED = 2
    SET_LED_FREQUENCY = 3


class ArduinoController:
    def __init__(self) -> None:
        self._transport: SerialTransport
        self._messenger: CmdMessenger
        self._form: WebForm

    def setup(self, form: WebForm) -> None:
        self._form = form

        # Register browser → Python commands.
        form.on_command("set_frequency", self._on_slider_change)
        form.on_command("set_led", self._on_toggle)

        self._transport = SerialTransport(
            SerialSettings(port_name=PORT, baud_rate=BAUD)
        )
        self._messenger = CmdMessenger(self._transport, board_type=BoardType.BIT_16)

        # Attach callbacks.
        self._messenger.attach(self._on_unknown)
        self._messenger.attach(Command.ACKNOWLEDGE, self._on_acknowledge)
        self._messenger.attach(Command.ERROR, self._on_error)

        self._messenger.new_line_received += lambda cmd: self._form.log_message(f"Received > {cmd.command_string()}")
        self._messenger.new_line_sent += lambda cmd: self._form.log_message(f"Sent > {cmd.command_string()}")

        if not self._messenger.connect():
            self._form.log_message(f"Could not open {PORT}")
            return

        # Initial state.
        self.set_led_state(True)
        self.set_led_frequency(2.0)

    def exit(self) -> None:
        self._messenger.disconnect()
        self._messenger.dispose()
        self._transport.dispose()

    # ------------------------------------------------------------------
    # Commands to Arduino
    # ------------------------------------------------------------------
    def set_led_frequency(self, frequency: float) -> None:
        command = SendCommand(Command.SET_LED_FREQUENCY, frequency)
        # CollapseCommandStrategy avoids queue lag when the slider fires rapidly.
        self._messenger.queue_command(CollapseCommandStrategy(command))

    def set_led_state(self, on: bool) -> None:
        self._messenger.send_command(SendCommand(Command.SET_LED, on))

    # ------------------------------------------------------------------
    # Browser → Python handlers
    # ------------------------------------------------------------------
    def _on_slider_change(self, value) -> None:
        try:
            self.set_led_frequency(float(value))
        except (TypeError, ValueError):
            pass

    def _on_toggle(self, value) -> None:
        self.set_led_state(bool(value))

    # ------------------------------------------------------------------
    # Arduino → Python callbacks
    # ------------------------------------------------------------------
    def _on_unknown(self, cmd: ReceivedCommand) -> None:
        self._form.log_message("Command without attached callback received")

    def _on_acknowledge(self, cmd: ReceivedCommand) -> None:
        self._form.log_message("Arduino is ready")

    def _on_error(self, cmd: ReceivedCommand) -> None:
        self._form.log_message("Arduino has experienced an error")

File: extras/CSharp/Samples/ArduinoJoystick/ArduinoController.cs
// *** ArduinoController ***

// This example expands the SendandReceiveArguments example. The PC will now sends commands to the Arduino when the trackbar 
// is pulled. Every TrackBarChanged events will queue a message to the Arduino to set the blink speed of the 
// internal / pin 13 LED
// 
// This example shows how to :
// - use in combination with WinForms
// - use in combination with ZedGraph
// - send queued commands
// - Use the CollapseCommandStrategy

using System;
using CommandMessenger;
using CommandMessenger.Queue;
using CommandMessenger.Transport.Serial;

namespace ArduinoController
{
    enum Command
    {
        Acknowledge,               // Command to acknowledge that cmd was received
        Error,                     // Command to report errors to host
        JoystickLocation,          // Command to send joystick position to host
        JoystickCenterCalibration, // Command to request joystick center calibration on device
    };

    public class ArduinoController
    {
        // This class (kind of) contains presentation logic, and domain model.
        // ChartForm.cs contains the view components 

        private SerialTransport   _serialTransport;
        private CmdMessenger      _cmdMessenger;
        private ControllerForm    _controllerForm;

        // ------------------ MAIN  ----------------------

        // Setup function
        public void Setup(ControllerForm controllerForm)
        {
            // storing the controller form for later reference
            _controllerForm = controllerForm;

            // Create Serial Port object
            // Note that for some boards (e.g. Sparkfun Pro Micro) DtrEnable may need to be true.
            _serialTransport = new SerialTransport
            {
                CurrentSerialSettings = { PortName = "COM10", BaudRate = 115200, DtrEnable = false } // object initializer
            };

            // Initialize the command messenger with the Serial Port transport layer
            // Set if it is communicating with a 16- or 32-bit Arduino board
            _cmdMessenger = new CmdMessenger(_serialTransport, BoardType.Bit16);

            // Tell CmdMessenger to "Invoke" commands on the thread running the WinForms UI
            _cmdMessenger.ControlToInvokeOn = _controllerForm;

            // Attach the callbacks to the Command Messenger
            AttachCommandCallBacks();

            // Attach to NewLinesReceived for logging purposes
            _cmdMessenger.NewLineReceived += NewLineReceived;

            // Attach to NewLineSent for logging purposes
            _cmdMessenger.NewLineSent += NewLineSent;                       

            // Start listening
            _cmdMessenger.Connect();

        }

        // Exit function
        public void Exit()
        {
            // Stop listening
            _cmdMessenger.Disconnect();

            // Dispose Command Messenger
            _cmdMessenger.Dispose();

            // Dispose Serial Port object
            _serialTransport.Dispose();
        }

        /// Attach command call backs. 
        private void AttachCommandCallBacks()
        {
            _cmdMessenger.Attach(OnUnknownCommand);
            _cmdMessenger.Attach((int)Command.Acknowledge, OnAcknowledge);
            _cmdMessenger.Attach((int)Command.Error, OnError);
            _cmdMessenger.Attach((int)Command.JoystickLocation, OnJoystickLocation);
        }

        // ------------------  CALLBACKS ---------------------

        // Called when a received command has no attached function.
        // In a WinForm application, console output gets routed to the output panel of your IDE
        void OnUnknownCommand(ReceivedCommand arguments)
        {            
            Console.WriteLine(@"Command without attached callback received");
        }

        // Callback function that prints that the Arduino has acknowledged
        void OnAcknowledge(ReceivedCommand arguments)
        {
            Console.WriteLine(@" Arduino is ready");
        }

        // Callback function that prints that the Arduino has experienced an error
        void OnError(ReceivedCommand arguments)
        {
            Console.WriteLine(@"Arduino has experienced an error");
        }

        // Log received line to console
        private void NewLineReceived(object sender, CommandEventArgs e)
        {
            Console.WriteLine(@"Received > " + e.Command.CommandString());
        }

        // Log sent line to console
        private void NewLineSent(object sender, CommandEventArgs e)
        {
            Console.WriteLine(@"Sent > " + e.Command.CommandString());
        }

        private void OnJoystickLocation(ReceivedCommand arguments)
        {

            // Get all arguments from plot data point command           
            var posY = arguments.ReadInt16Arg();
            var posX = arguments.ReadInt16Arg();

            _controllerForm.trackBarX.Value = posX;
            _controllerForm.trackBarY.Value = posY;

        }


    }
}
